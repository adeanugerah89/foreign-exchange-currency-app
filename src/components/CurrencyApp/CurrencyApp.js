import React, { Component } from 'react';
import axios from 'axios';
import CurrencyListApp from '../CurrencyListApp/CurrencyListApp';
import { Button, FormControl } from 'react-bootstrap';
import '../../styles/CurrencyAppCss.css';
import '../../styles/CurrencyListAppCss.css';

class CurrencyApp extends Component {
    constructor(props) {
        super(props);
        this.state = {
            addCurrencies: '',
            dataList: [
                {
                    baseValue: "14901.50",
                    id: 1538372038451,
                    key: "IDR",
                    value: "14901.50",
                },
                {
                    baseValue: "0.86",
                    id: 1538372238662,
                    key: "EUR",
                    value: 0.86
                },
                {
                    baseValue: "0.77",
                    id: 1538372337662,
                    key: "GBP",
                    value: 0.77
                },
                {
                    baseValue: "1.37",
                    id: 1538372397202,
                    key: "SGD",
                    value: 1.37
                }
            ],
            dataSource: []
        };

        this.handleAddCurrencies = this.handleAddCurrencies.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.deleteItem = this.deleteItem.bind(this);
    }

    componentDidMount() {
        axios.get(`https://api.exchangeratesapi.io/latest?base=USD`)
          .then((response) => {
              let dataRates = response.data
              for(let value in dataRates.rates){
                  dataRates.rates[value] = dataRates.rates[value].toFixed(2)
                }

              this.setState({
                dataSource: dataRates
              })
          })
          .catch((error) => console.log(error))
     }

    componentDidUpdate (prevProps){
        if(this.props !== prevProps){
            let newStateValue = this.state.dataList.map(data => data.baseValue * this.props.inputValue)
            let newDataList = []
            this.state.dataList.forEach((data, idx) => {
                 newDataList.push({
                     key: data.key,
                     value:  newStateValue[idx],
                     baseValue:data.baseValue,
                     id: data.id
                 })
            })

            this.setState({
                dataList: newDataList
            })
        }

        return true
    }

    deleteItem(id) {
        let filteredItems = this.state.dataList.filter((item) => {
            return (item.id !== id)
        });

        this.setState({
            dataList: filteredItems
        })
    }

    handleAddCurrencies(event) {
        this.setState({addCurrencies: event.target.value});
    }

    handleSubmit(event) {
        let dataSource = this.state.dataSource.rates
        let addCurrencies = this.state.addCurrencies.toUpperCase()
        let existingKey = this.state.dataList.filter(data => data.key === addCurrencies)
        if (event.target.value !== "" && existingKey.length === 0){
            let newAddDataListItem = {}
            for (let key in dataSource) {
                if (key === addCurrencies) {
                    let value = this.props.inputValue ? this.props.inputValue * dataSource[key] : dataSource[key]
                    newAddDataListItem = {
                        key: key,
                        value: value,
                        baseValue: dataSource[key],
                        id: Date.now()
                    };
                }
            }

            if(Object.keys(newAddDataListItem).length !== 0) {
                this.setState((prev) => {
                    return { dataList: prev.dataList.concat(newAddDataListItem) }
                })
            }else {
                alert('Data not found, please check your input text')
            }

        }
        event.preventDefault();
        this.setState({addCurrencies: ''});
     }

    render(){
        return (
            <div className="currencyListMain">
                <CurrencyListApp
                    entries={this.state.dataList}
                    delete={this.deleteItem}
                />
                <div className="header">
                    <form onSubmit={this.handleSubmit}>
                        <FormControl
                          type="text"
                          value={this.state.addCurrencies}
                          placeholder="add more currencies"
                          onChange={this.handleAddCurrencies}
                        />
                        <Button bsStyle="primary" type="submit">Submit</Button>
                    </form>
                </div>
            </div>
        )
    }
}

export default CurrencyApp;
