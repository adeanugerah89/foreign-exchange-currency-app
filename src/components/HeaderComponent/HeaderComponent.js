import React, { Component } from 'react';
import { Panel, FormControl, FormGroup, ControlLabel, InputGroup } from 'react-bootstrap';
import CurrencyApp from '../CurrencyApp/CurrencyApp';

class HeaderComponent extends Component {
    constructor(props) {
        super(props);
        this.state = {
            inputValue: '',
        }

        this.handleInputValue = this.handleInputValue.bind(this);
    }

    handleInputValue(event) {
        this.setState({inputValue: Number(event.target.value) ? Number(event.target.value) : ''})
    }

    render() {
        return (
            <div style={{width: 400}}>
                <Panel bsStyle="info">
                    <Panel.Heading>
                      <Panel.Title componentClass="h3">
                          <FormGroup>
                            <ControlLabel>USD - United State Dollars</ControlLabel>
                            <InputGroup>
                              <InputGroup.Addon>$</InputGroup.Addon>
                              <FormControl
                                pattern='[0-9]'
                                type="text"
                                value={this.state.inputValue}
                                placeholder="Enter value with number format"
                                onChange={this.handleInputValue}
                              />
                            </InputGroup>
                          </FormGroup>
                      </Panel.Title>
                    </Panel.Heading>
                    <Panel.Body>
                        <CurrencyApp inputValue={this.state.inputValue}/>
                    </Panel.Body>
                </Panel>
            </div>
        )
    }
}

export default HeaderComponent;
