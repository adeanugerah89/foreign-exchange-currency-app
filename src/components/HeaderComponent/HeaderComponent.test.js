import React from 'react';
import {mount, render, shallow} from 'enzyme'
import '../setupTest'

import HeaderComponent from './HeaderComponent';
import CurrencyApp from '../CurrencyApp/CurrencyApp';

describe('<HeaderComponent /> test', () => {

    it('renders correctly', () => {
        const wrapper = shallow(<HeaderComponent />);
        expect(wrapper).toMatchSnapshot();
    });

    it('HeaderComponent should have state', () => {
        const wrapper = shallow(<HeaderComponent />);
        expect(wrapper.state('inputValue')).toEqual('');
    });

    it('renders one <CurrencyApp /> components', () => {
        const wrapper = shallow(<HeaderComponent />);
        expect(wrapper.find(CurrencyApp)).toHaveLength(1);
    });

    it('should have text input', () => {
        const wrapper = shallow(<HeaderComponent />);
        expect(wrapper.find('FormControl')).toHaveLength(1);
    });

});
