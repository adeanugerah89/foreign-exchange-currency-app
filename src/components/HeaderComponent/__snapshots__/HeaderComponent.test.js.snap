// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeaderComponent /> test renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <HeaderComponent />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Uncontrolled(Panel)
        bsStyle="info"
>
        <PanelHeading
                bsClass="panel"
                componentClass="div"
        >
                <PanelTitle
                        bsClass="panel"
                        componentClass="h3"
                >
                        <FormGroup
                                bsClass="form-group"
                        >
                                <ControlLabel
                                        bsClass="control-label"
                                        srOnly={false}
                                >
                                        USD - United State Dollars
                                </ControlLabel>
                                <InputGroup
                                        bsClass="input-group"
                                >
                                        <InputGroupAddon
                                                bsClass="input-group-addon"
                                        >
                                                $
                                        </InputGroupAddon>
                                        <FormControl
                                                bsClass="form-control"
                                                componentClass="input"
                                                onChange={[Function]}
                                                pattern="[0-9]"
                                                placeholder="Enter value with number format"
                                                type="text"
                                                value=""
                                        />
                                </InputGroup>
                        </FormGroup>
                </PanelTitle>
        </PanelHeading>
        <PanelBody
                bsClass="panel"
                collapsible={false}
        >
                <CurrencyApp
                        inputValue=""
                />
        </PanelBody>
</Uncontrolled(Panel)>,
      "style": Object {
        "width": 400,
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "bsStyle": "info",
        "children": Array [
          <PanelHeading
            bsClass="panel"
            componentClass="div"
>
            <PanelTitle
                        bsClass="panel"
                        componentClass="h3"
            >
                        <FormGroup
                                    bsClass="form-group"
                        >
                                    <ControlLabel
                                                bsClass="control-label"
                                                srOnly={false}
                                    >
                                                USD - United State Dollars
                                    </ControlLabel>
                                    <InputGroup
                                                bsClass="input-group"
                                    >
                                                <InputGroupAddon
                                                            bsClass="input-group-addon"
                                                >
                                                            $
                                                </InputGroupAddon>
                                                <FormControl
                                                            bsClass="form-control"
                                                            componentClass="input"
                                                            onChange={[Function]}
                                                            pattern="[0-9]"
                                                            placeholder="Enter value with number format"
                                                            type="text"
                                                            value=""
                                                />
                                    </InputGroup>
                        </FormGroup>
            </PanelTitle>
</PanelHeading>,
          <PanelBody
            bsClass="panel"
            collapsible={false}
>
            <CurrencyApp
                        inputValue=""
            />
</PanelBody>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "panel",
            "children": <PanelTitle
              bsClass="panel"
              componentClass="h3"
>
              <FormGroup
                            bsClass="form-group"
              >
                            <ControlLabel
                                          bsClass="control-label"
                                          srOnly={false}
                            >
                                          USD - United State Dollars
                            </ControlLabel>
                            <InputGroup
                                          bsClass="input-group"
                            >
                                          <InputGroupAddon
                                                        bsClass="input-group-addon"
                                          >
                                                        $
                                          </InputGroupAddon>
                                          <FormControl
                                                        bsClass="form-control"
                                                        componentClass="input"
                                                        onChange={[Function]}
                                                        pattern="[0-9]"
                                                        placeholder="Enter value with number format"
                                                        type="text"
                                                        value=""
                                          />
                            </InputGroup>
              </FormGroup>
</PanelTitle>,
            "componentClass": "div",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "panel",
              "children": <FormGroup
                bsClass="form-group"
>
                <ControlLabel
                                bsClass="control-label"
                                srOnly={false}
                >
                                USD - United State Dollars
                </ControlLabel>
                <InputGroup
                                bsClass="input-group"
                >
                                <InputGroupAddon
                                                bsClass="input-group-addon"
                                >
                                                $
                                </InputGroupAddon>
                                <FormControl
                                                bsClass="form-control"
                                                componentClass="input"
                                                onChange={[Function]}
                                                pattern="[0-9]"
                                                placeholder="Enter value with number format"
                                                type="text"
                                                value=""
                                />
                </InputGroup>
</FormGroup>,
              "componentClass": "h3",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "form-group",
                "children": Array [
                  <ControlLabel
                    bsClass="control-label"
                    srOnly={false}
>
                    USD - United State Dollars
</ControlLabel>,
                  <InputGroup
                    bsClass="input-group"
>
                    <InputGroupAddon
                                        bsClass="input-group-addon"
                    >
                                        $
                    </InputGroupAddon>
                    <FormControl
                                        bsClass="form-control"
                                        componentClass="input"
                                        onChange={[Function]}
                                        pattern="[0-9]"
                                        placeholder="Enter value with number format"
                                        type="text"
                                        value=""
                    />
</InputGroup>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "bsClass": "control-label",
                    "children": "USD - United State Dollars",
                    "srOnly": false,
                  },
                  "ref": null,
                  "rendered": "USD - United State Dollars",
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "bsClass": "input-group",
                    "children": Array [
                      <InputGroupAddon
                        bsClass="input-group-addon"
>
                        $
</InputGroupAddon>,
                      <FormControl
                        bsClass="form-control"
                        componentClass="input"
                        onChange={[Function]}
                        pattern="[0-9]"
                        placeholder="Enter value with number format"
                        type="text"
                        value=""
/>,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "bsClass": "input-group-addon",
                        "children": "$",
                      },
                      "ref": null,
                      "rendered": "$",
                      "type": [Function],
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "bsClass": "form-control",
                        "componentClass": "input",
                        "onChange": [Function],
                        "pattern": "[0-9]",
                        "placeholder": "Enter value with number format",
                        "type": "text",
                        "value": "",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                  ],
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "panel",
            "children": <CurrencyApp
              inputValue=""
/>,
            "collapsible": false,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "inputValue": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Uncontrolled(Panel)
          bsStyle="info"
>
          <PanelHeading
                    bsClass="panel"
                    componentClass="div"
          >
                    <PanelTitle
                              bsClass="panel"
                              componentClass="h3"
                    >
                              <FormGroup
                                        bsClass="form-group"
                              >
                                        <ControlLabel
                                                  bsClass="control-label"
                                                  srOnly={false}
                                        >
                                                  USD - United State Dollars
                                        </ControlLabel>
                                        <InputGroup
                                                  bsClass="input-group"
                                        >
                                                  <InputGroupAddon
                                                            bsClass="input-group-addon"
                                                  >
                                                            $
                                                  </InputGroupAddon>
                                                  <FormControl
                                                            bsClass="form-control"
                                                            componentClass="input"
                                                            onChange={[Function]}
                                                            pattern="[0-9]"
                                                            placeholder="Enter value with number format"
                                                            type="text"
                                                            value=""
                                                  />
                                        </InputGroup>
                              </FormGroup>
                    </PanelTitle>
          </PanelHeading>
          <PanelBody
                    bsClass="panel"
                    collapsible={false}
          >
                    <CurrencyApp
                              inputValue=""
                    />
          </PanelBody>
</Uncontrolled(Panel)>,
        "style": Object {
          "width": 400,
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "bsStyle": "info",
          "children": Array [
            <PanelHeading
              bsClass="panel"
              componentClass="div"
>
              <PanelTitle
                            bsClass="panel"
                            componentClass="h3"
              >
                            <FormGroup
                                          bsClass="form-group"
                            >
                                          <ControlLabel
                                                        bsClass="control-label"
                                                        srOnly={false}
                                          >
                                                        USD - United State Dollars
                                          </ControlLabel>
                                          <InputGroup
                                                        bsClass="input-group"
                                          >
                                                        <InputGroupAddon
                                                                      bsClass="input-group-addon"
                                                        >
                                                                      $
                                                        </InputGroupAddon>
                                                        <FormControl
                                                                      bsClass="form-control"
                                                                      componentClass="input"
                                                                      onChange={[Function]}
                                                                      pattern="[0-9]"
                                                                      placeholder="Enter value with number format"
                                                                      type="text"
                                                                      value=""
                                                        />
                                          </InputGroup>
                            </FormGroup>
              </PanelTitle>
</PanelHeading>,
            <PanelBody
              bsClass="panel"
              collapsible={false}
>
              <CurrencyApp
                            inputValue=""
              />
</PanelBody>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "panel",
              "children": <PanelTitle
                bsClass="panel"
                componentClass="h3"
>
                <FormGroup
                                bsClass="form-group"
                >
                                <ControlLabel
                                                bsClass="control-label"
                                                srOnly={false}
                                >
                                                USD - United State Dollars
                                </ControlLabel>
                                <InputGroup
                                                bsClass="input-group"
                                >
                                                <InputGroupAddon
                                                                bsClass="input-group-addon"
                                                >
                                                                $
                                                </InputGroupAddon>
                                                <FormControl
                                                                bsClass="form-control"
                                                                componentClass="input"
                                                                onChange={[Function]}
                                                                pattern="[0-9]"
                                                                placeholder="Enter value with number format"
                                                                type="text"
                                                                value=""
                                                />
                                </InputGroup>
                </FormGroup>
</PanelTitle>,
              "componentClass": "div",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "panel",
                "children": <FormGroup
                  bsClass="form-group"
>
                  <ControlLabel
                                    bsClass="control-label"
                                    srOnly={false}
                  >
                                    USD - United State Dollars
                  </ControlLabel>
                  <InputGroup
                                    bsClass="input-group"
                  >
                                    <InputGroupAddon
                                                      bsClass="input-group-addon"
                                    >
                                                      $
                                    </InputGroupAddon>
                                    <FormControl
                                                      bsClass="form-control"
                                                      componentClass="input"
                                                      onChange={[Function]}
                                                      pattern="[0-9]"
                                                      placeholder="Enter value with number format"
                                                      type="text"
                                                      value=""
                                    />
                  </InputGroup>
</FormGroup>,
                "componentClass": "h3",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "form-group",
                  "children": Array [
                    <ControlLabel
                      bsClass="control-label"
                      srOnly={false}
>
                      USD - United State Dollars
</ControlLabel>,
                    <InputGroup
                      bsClass="input-group"
>
                      <InputGroupAddon
                                            bsClass="input-group-addon"
                      >
                                            $
                      </InputGroupAddon>
                      <FormControl
                                            bsClass="form-control"
                                            componentClass="input"
                                            onChange={[Function]}
                                            pattern="[0-9]"
                                            placeholder="Enter value with number format"
                                            type="text"
                                            value=""
                      />
</InputGroup>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "bsClass": "control-label",
                      "children": "USD - United State Dollars",
                      "srOnly": false,
                    },
                    "ref": null,
                    "rendered": "USD - United State Dollars",
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "bsClass": "input-group",
                      "children": Array [
                        <InputGroupAddon
                          bsClass="input-group-addon"
>
                          $
</InputGroupAddon>,
                        <FormControl
                          bsClass="form-control"
                          componentClass="input"
                          onChange={[Function]}
                          pattern="[0-9]"
                          placeholder="Enter value with number format"
                          type="text"
                          value=""
/>,
                      ],
                    },
                    "ref": null,
                    "rendered": Array [
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "class",
                        "props": Object {
                          "bsClass": "input-group-addon",
                          "children": "$",
                        },
                        "ref": null,
                        "rendered": "$",
                        "type": [Function],
                      },
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "class",
                        "props": Object {
                          "bsClass": "form-control",
                          "componentClass": "input",
                          "onChange": [Function],
                          "pattern": "[0-9]",
                          "placeholder": "Enter value with number format",
                          "type": "text",
                          "value": "",
                        },
                        "ref": null,
                        "rendered": null,
                        "type": [Function],
                      },
                    ],
                    "type": [Function],
                  },
                ],
                "type": [Function],
              },
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "panel",
              "children": <CurrencyApp
                inputValue=""
/>,
              "collapsible": false,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "inputValue": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
